src/Typin/Typin/Experimental/Design.cs

```diff
@@ -0,0 +1,174 @@
+    {
+        /// <summary>
+        /// Creates and returns a new instance of <see cref="ICommandRunner"/> capable of executing
+        /// the task this type represents.
+        /// </summary>
+        /// <returns></returns>
+        ICommandRunner CreateRunner();
+    }
+
+    /// <summary>
+    /// Describes a <see cref="ICommandType"/> with human readable information and command input specifications
+    /// </summary>
+    public class CommandDescriptor
+    {
+        /// <summary>
+        /// Default name to invoke the command. It can be a space separated collection of tokens to create
+        /// a command hierarchy, like <code>pods</code> and <code>pods list all</code>
+        /// </summary>
+        public string Name { get; init; }
+
+        /// <summary>
+        /// A short summary able to be printed in a single console line, when the command is presented 
+        /// in a list of other sibling commands
+        /// </summary>
+        public string Summary { get; init; }
+
+        /// <summary>
+        /// A description, possibly more extense than the <see cref="Summary"/> decribing what the command does
+        /// </summary>
+        public string Description { get; init; }
+
+        /// <summary>
+        /// List of argument descriptors specifying the arguments of current command
+        /// </summary>
+        public IReadOnlyCollection<CommandArgumentDescriptor> Arguments { get; init; }
+
+        /// <summary>
+        /// List of option descriptors specifying the options of current command
+        /// </summary>
+        public IReadOnlyCollection<CommandOptionDescriptor> Options { get; init; }
+
+        /// <summary>
+        /// List of examples to show to the user for current command
+        /// </summary>
+        public IReadOnlyCollection<CommandExampleDescriptor> Examples { get; init; }
+    }
+
+    public class CommandArgumentDescriptor
+    {
+        public string Name { get; init; }
+        public string Summary { get; init; }
+        public string Description { get; init; }
+        public int Position { get; init; }
+        public bool Multiple { get; init; }
+    }
+
+    public enum OptionKind
+    {
+        Flag,
+        Value,
+    }
+
+    public class CommandOptionDescriptor
+    {
+        public string Name { get; init; }
+        public IReadOnlyCollection<string> ShortAliases { get; init; }
+
+        /// <summary>
+        /// For flags, you can specify a negative name, like no-verbose, no-build, etc., so the user can
+        /// explicitly indicate that he wants to give a false value for the flag, disregarding it's default value
+        /// </summary>
+        public string NegativeName { get; init; }
+        public string Summary { get; init; }
+        public string Description { get; init; }
+        public bool Multiple { get; init; }
+        public OptionKind OptionKind { get; init; }
+        public string DefaultValue { get; init; }
```

Comment:
Agree, but beyond that, I would like the parameter/option to have a way to parse strings into corresponding objects. Something like:

```csharp
public bool TryParseValueDelegate(string text, out object value);

public class CommandOptionDescriptor
{
	public TryParseValueDelegate TryParseValue { get; init; }
}
```

Taking [this](https://github.com/adambajguz/Typin/wiki/Value-conversion) as current I cannot say we have any extensibility for custom types other than what you defined.

Maybe instead of a Delegate we can have an actual interface.

Then, indeed, `DefaultValue` can be an object, but I don't know why to distinguish between dynamic and non-dynamic command (I don't do it in my head).

I do not understand how a default value is not constant? You mean like in C# constant? I'm trying to separate the schemas from the C# types and properties. There will be a core implementation of `ICommandProvider` that will implement `ICommand` definitions using Reflection as currently, but that will be one of many, not the only one option as currently.

Reply:
_"Taking this as current I cannot say we have any extensibility for custom types other than what you defined."_ - yes but we #7 will fix this problem (also see: #203 )

"I do not understand how a default value is not constant?" - simple example:
```c#
    [Command]
    public class NonConstDefaultCommand : ICommand
    {
        [CommandOption]
        public DateTime Date { get; init; } = DateTime.Now;

        public ValueTask ExecuteAsync(IConsole console)
        {
            console.Output.WriteLine(Date.ToString());

            return default;
        }
    }
```